File: modules/core/src/main/java/com/panayotis/jubler/options/Options.java
/*
 * (c) 2005-2025 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

package com.panayotis.jubler.options;

import com.panayotis.appenh.EnhancerManager;
import com.panayotis.appenh.ThemeVariation;
import com.panayotis.jubler.JublerPrefs;
import com.panayotis.jubler.options.gui.TabPage;
import com.panayotis.jubler.subs.SubFile;

import java.io.File;
import java.nio.charset.Charset;
import java.util.Stack;

@SuppressWarnings("UseSpecificCatch")
public class Options {

    public final static int CURRENT_VERSION = 3;
    private final static int MAX_RECENTS = 10;
    private static int errorColor;
    private static final String ERRORCOLOR_TAG = "options.errorcolor";
    private static boolean spaceChars;
    private static final String SPACECHARS_TAG = "options.spaceaschars";
    private static boolean newlineChars;
    private static final String NEWLINECHARS_TAG = "options.newlineaschars";
    private static boolean compactSubs;
    private static final String COMPACTSUBS_TAG = "options.compactsubs";
    private static boolean otherChars;
    private static final String OTHERCHARS_TAG = "options.otheraschars";
    private static int maxLines;
    private static final String MAXLINES_TAG = "options.maxline";
    private static float fillPercent;
    private static final String FILLPERCENT_TAG = "options.fillpercent";
    private static int maxLineLength;
    private static final String MAXLINELENGTH_TAG = "options.maxlinelength";
    private static int maxCPS;
    private static final String MAXCPS_TAG = "options.maxcps";
    private static float maxDuration;
    private static final String MAXDURATION_TAG = "options.maxduration";
    private static float minDuration;
    private static final String MINDURATION_TAG = "options.minduration";
    private static final String SYSTEM_LASTFILE = "system.lastfile";
    private static float scaling;
    private static final String SCALING_TAG = "ui.scaling.factor";
    private static boolean timestampTooltipsDisabled;
    private static final String TIMESTAMP_TOOLTIPS_DISABLED = "ui.tooltips.timestamp.disabled";
    private static ThemeVariation themeVariation;
    private static final String USE_THEME_VARIATION = "ui.theme.variation";
    private static String language;
    private static final String LANGUAGE_TAG = "ui.language";
    private static int audioCacheRate;
    private static final String AUDIOCACHE_RATE_TAG = "audiocache.rate";
    private static int audioCacheChannels;
    private static final String AUDIOCACHE_CHANNELS_TAG = "audiocache.channels";
    private static boolean audioCacheDeleteOnClose;
    private static final String AUDIOCACHE_DELETEONCLOSE_TAG = "audiocache.deleteonclose";
    private static boolean videoPreviewHardware;
    private static final String VIDEOPREVIEW_HARDWARE_TAG = "videopreview.hardware";
    private static String defaultEncoding8bit;
    private static final String DEFAULT_ENCODING_8BIT_TAG = "default.encoding.8bit";
    private static String defaultEncodingCjk;
    private static final String DEFAULT_ENCODING_CJK_TAG = "default.encoding.cjk";

    public static final int AUDIOCACHE_DEFAULT_RATE = 16000;
    public static final int AUDIOCACHE_DEFAULT_CHANNELS = 2;

    static {
        errorColor = JublerPrefs.getInt(ERRORCOLOR_TAG, 1);
        spaceChars = JublerPrefs.getBoolean(SPACECHARS_TAG, false);
        newlineChars = JublerPrefs.getBoolean(NEWLINECHARS_TAG, false);
        compactSubs = JublerPrefs.getBoolean(COMPACTSUBS_TAG, true);
        otherChars = JublerPrefs.getBoolean(OTHERCHARS_TAG, true);
        maxLines = JublerPrefs.getInt(MAXLINES_TAG, 2);
        fillPercent = JublerPrefs.getFloat(FILLPERCENT_TAG, 50);
        maxLineLength = JublerPrefs.getInt(MAXLINELENGTH_TAG, 42);
        maxCPS = JublerPrefs.getInt(MAXCPS_TAG, 21);
        maxDuration = JublerPrefs.getFloat(MAXDURATION_TAG, 7);
        minDuration = JublerPrefs.getFloat(MINDURATION_TAG, 1);
        scaling = JublerPrefs.getFloat(SCALING_TAG, Math.max(1f, EnhancerManager.getDefault().getDPI() / 96f));
        timestampTooltipsDisabled = JublerPrefs.getBoolean(TIMESTAMP_TOOLTIPS_DISABLED, false);
        String theme = JublerPrefs.getString(USE_THEME_VARIATION, ThemeVariation.AUTO.name());
        try {
            themeVariation = ThemeVariation.valueOf(theme);
        } catch (Exception e) {
            themeVariation = ThemeVariation.AUTO;
        }
        language = JublerPrefs.getString(LANGUAGE_TAG, "auto");
        audioCacheRate = JublerPrefs.getInt(AUDIOCACHE_RATE_TAG, AUDIOCACHE_DEFAULT_RATE);
        audioCacheChannels = JublerPrefs.getInt(AUDIOCACHE_CHANNELS_TAG, AUDIOCACHE_DEFAULT_CHANNELS);
        audioCacheDeleteOnClose = JublerPrefs.getBoolean(AUDIOCACHE_DELETEONCLOSE_TAG, false);
        videoPreviewHardware = JublerPrefs.getBoolean(VIDEOPREVIEW_HARDWARE_TAG, false);
    }

    static {
        migrateDefaultEncoding();   // must run before the defaults are read below
        defaultEncoding8bit = JublerPrefs.getString(DEFAULT_ENCODING_8BIT_TAG, "ISO-8859-1");
        if (!isSingleByteCharset(defaultEncoding8bit))   // the floor must always decode → single-byte
            defaultEncoding8bit = "ISO-8859-1";
        defaultEncodingCjk = JublerPrefs.getString(DEFAULT_ENCODING_CJK_TAG, null);
//        JublerPrefs.dump();
    }

    /**
     * True for the self-identifying Unicode charsets (UTF-8/16/32) — found automatically (UTF-8 by
     * strict validation, UTF-16/32 by BOM), so they are never remembered.
     */
    public static boolean isUnicodeCharset(String name) {
        if (name == null)
            return false;
        String u = name.toUpperCase();
        return u.startsWith("UTF-") || u.startsWith("UTF8") || u.startsWith("UTF16")
                || u.startsWith("UTF32") || u.startsWith("X-UTF") || u.equals("UTF");
    }

    /** True for genuine single-byte charsets (one byte per character) — they always decode any bytes. */
    public static boolean isSingleByteCharset(String name) {
        try {
            return Charset.forName(name).newEncoder().maxBytesPerChar() == 1f;
        } catch (Exception e) {
            return false;
        }
    }

    /** The remembered single-byte charset — the always-succeeding floor of the load-time auto-detect. */
    public static String getDefaultEncoding8bit() {
        return defaultEncoding8bit;
    }

    /** The remembered multi-byte CJK charset, or null if none — tried (strict) before the floor. */
    public static String getDefaultEncodingCjk() {
        return defaultEncodingCjk;
    }

    /**
     * Remember a chosen charset and persist it, routed to the right slot: a genuine single-byte
     * charset becomes the floor, a multi-byte CJK charset becomes the CJK hint. Unicode is ignored
     * (self-identifying) — this is what makes the bar's persistence conditional: picking UTF is
     * transient, picking any non-Unicode charset sticks in its slot.
     */
    public static void rememberEncoding(String enc) {
        if (enc == null || enc.isEmpty() || isUnicodeCharset(enc))
            return;
        if (isSingleByteCharset(enc)) {
            defaultEncoding8bit = enc;
            JublerPrefs.set(DEFAULT_ENCODING_8BIT_TAG, enc);
        } else {
            defaultEncodingCjk = enc;
            JublerPrefs.set(DEFAULT_ENCODING_CJK_TAG, enc);
        }
    }

    /**
     * One-time preferences migration: split the legacy {@code default.encoding1/2/3} triplet into the
     * two slots — first single-byte entry → floor (else ISO-8859-1), first multi-byte CJK entry → CJK
     * hint (if any). Idempotent; the old keys are left untouched so a downgrade still works.
     */
    private static void migrateDefaultEncoding() {
        if (JublerPrefs.getString(DEFAULT_ENCODING_8BIT_TAG, null) != null)
            return;
        String single = null, cjk = null;
        for (int i = 1; i <= 3; i++) {
            String e = JublerPrefs.getString("default.encoding" + i, null);
            if (e == null || isUnicodeCharset(e))
                continue;
            if (isSingleByteCharset(e)) {
                if (single == null)
                    single = e;
            } else if (cjk == null)
                cjk = e;
        }
        JublerPrefs.set(DEFAULT_ENCODING_8BIT_TAG, single == null ? "ISO-8859-1" : single);
        if (cjk != null)
            JublerPrefs.set(DEFAULT_ENCODING_CJK_TAG, cjk);
    }

    public static void loadSystemPreferences(JPreferences prefs) {
        for (TabPage opt : prefs.Tabs.getTabArray())
            ((OptionsHolder) opt).loadPreferences();
    }

    public static void saveSystemPreferences(JPreferences prefs) {
        for (TabPage opt : prefs.Tabs.getTabArray())
            ((OptionsHolder) opt).savePreferences();
    }

    public static void saveFileList(Stack<SubFile> recents) {
        SubFile sfile;
        File f;
        int pos = recents.size();
        int counter = 0;
        while (pos > 0 && counter < MAX_RECENTS) {
            pos--;
            sfile = recents.get(pos);
            f = sfile.getSaveFile();
            if (f.exists() && f.isFile()) {
                counter++;
                JublerPrefs.set(SYSTEM_LASTFILE + counter, sfile.getPacked());
            }
        }
        while (counter < MAX_RECENTS)
            JublerPrefs.set(SYSTEM_LASTFILE + (++counter), null);
    }

    public static Stack<SubFile> loadFileList() {
        Stack<SubFile> files = new Stack<SubFile>();
        File f;
        for (int i = MAX_RECENTS; i > 0; i--)
            try {
                SubFile sf = new SubFile(JublerPrefs.getString(SYSTEM_LASTFILE + i, ""));
                f = sf.getSaveFile();
                if (f.exists() && f.canRead() && f.isFile())
                    files.push(sf);
            } catch (InstantiationException ignored) {
            }
        return files;
    }

    public static float getScaling() {
        return scaling;
    }

    public static void setScaling(float newScaling) {
        JublerPrefs.set(SCALING_TAG, newScaling);
        scaling = newScaling;
    }

    public static boolean isTimestampTooltipsDisabled() {
        return timestampTooltipsDisabled;
    }

    public static void setTimestampTooltipsDisabled(boolean disabled) {
        JublerPrefs.set(TIMESTAMP_TOOLTIPS_DISABLED, disabled);
        timestampTooltipsDisabled = disabled;
    }

    public static ThemeVariation getThemeVariation() {
        return themeVariation;
    }

    public static void setThemeVariation(ThemeVariation variation) {
        JublerPrefs.set(USE_THEME_VARIATION, variation.name());
        themeVariation = variation;
    }

    public static int getAudioCacheRate() {
        return audioCacheRate;
    }

    public static void setAudioCacheRate(int rate) {
        JublerPrefs.set(AUDIOCACHE_RATE_TAG, audioCacheRate = rate);
    }

    public static int getAudioCacheChannels() {
        return audioCacheChannels;
    }

    public static void setAudioCacheChannels(int channels) {
        JublerPrefs.set(AUDIOCACHE_CHANNELS_TAG, audioCacheChannels = channels);
    }

    public static boolean isAudioCacheDeleteOnClose() {
        return audioCacheDeleteOnClose;
    }

    public static void setAudioCacheDeleteOnClose(boolean delete) {
        JublerPrefs.set(AUDIOCACHE_DELETEONCLOSE_TAG, audioCacheDeleteOnClose = delete);
    }

    public static boolean isVideoPreviewHardware() {
        return videoPreviewHardware;
    }

    public static void setVideoPreviewHardware(boolean hardware) {
        JublerPrefs.set(VIDEOPREVIEW_HARDWARE_TAG, videoPreviewHardware = hardware);
    }

    public static String getLanguage() {
        return language;
    }

    public static void setLanguage(String lang) {
        JublerPrefs.set(LANGUAGE_TAG, lang);
        language = lang;
    }

    public static void setErrorColor(int newcolor) {
        JublerPrefs.set(ERRORCOLOR_TAG, errorColor = newcolor);
    }

    public static int getErrorColor() {
        return errorColor;
    }

    public static void setSpaceChars(boolean spaceAsChars) {
        JublerPrefs.set(SPACECHARS_TAG, spaceChars = spaceAsChars);
    }

    public static boolean isSpaceChars() {
        return spaceChars;
    }

    public static void setNewlineChars(boolean newlineAsChars) {
        JublerPrefs.set(NEWLINECHARS_TAG, newlineChars = newlineAsChars);
    }

    public static boolean isNewlineChars() {
        return newlineChars;
    }

    public static void setCompactSubs(boolean csubs) {
        JublerPrefs.set(COMPACTSUBS_TAG, compactSubs = csubs);
    }

    public static boolean isCompactSubs() {
        return compactSubs;
    }

    public static void setOtherChars(boolean otherAsChars) {
        JublerPrefs.set(OTHERCHARS_TAG, otherChars = otherAsChars);
    }

    public static boolean isOtherChars() {
        return otherChars;
    }

    public static void setMaxLines(int newmaxlines) {
        JublerPrefs.set(MAXLINES_TAG, maxLines = newmaxlines);
    }

    public static int getMaxLines() {
        return maxLines;
    }

    public static void setFillPercent(float value) {
        if (value < 0)
            value = 0;
        else if (value > 100)
            value = 100;
        JublerPrefs.set(FILLPERCENT_TAG, fillPercent = value);
    }

    public static float getFillPercent() {
        return fillPercent;
    }

    public static void setMaxLineLength(int maxlinelength) {
        JublerPrefs.set(MAXLINELENGTH_TAG, maxLineLength = maxlinelength);
    }

    public static int getMaxLineLength() {
        return maxLineLength;
    }

    public static void setMaxCPS(int maxcharssecond) {
        JublerPrefs.set(MAXCPS_TAG, maxCPS = maxcharssecond);
    }

    public static int getMaxCPS() {
        return maxCPS;
    }

    public static void setMaxDuration(float maxduration) {
        JublerPrefs.set(MAXDURATION_TAG, maxDuration = maxduration);
    }

    public static float getMaxDuration() {
        return maxDuration;
    }

    public static void setMinDuration(float minduration) {
        JublerPrefs.set(MAXDURATION_TAG, minDuration = minduration);
    }

    public static float getMinDuration() {
        return minDuration;
    }

}

File: src/core/options/Options.cpp
/*
 * (c) 2005-2026 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

#include "core/options/Options.h"

#include <QFileInfo>
#include "core/os/Charsets.h"

#include "core/options/Prefs.h"
#include "core/subs/SubFile.h"

QString themeVariationName(ThemeVariation v) {
    switch (v) {
        case ThemeVariation::LIGHT: return QStringLiteral("LIGHT");
        case ThemeVariation::DARK: return QStringLiteral("DARK");
        default: return QStringLiteral("AUTO");
    }
}

ThemeVariation themeVariationFromName(const QString &name) {
    if (name == QLatin1String("LIGHT")) return ThemeVariation::LIGHT;
    if (name == QLatin1String("DARK")) return ThemeVariation::DARK;
    return ThemeVariation::AUTO;
}

namespace Options {

namespace {
const QString ERRORCOLOR_TAG = QStringLiteral("options.errorcolor");
const QString SPACECHARS_TAG = QStringLiteral("options.spaceaschars");
const QString NEWLINECHARS_TAG = QStringLiteral("options.newlineaschars");
const QString COMPACTSUBS_TAG = QStringLiteral("options.compactsubs");
const QString OTHERCHARS_TAG = QStringLiteral("options.otheraschars");
const QString MAXLINES_TAG = QStringLiteral("options.maxline");
const QString FILLPERCENT_TAG = QStringLiteral("options.fillpercent");
const QString MAXLINELENGTH_TAG = QStringLiteral("options.maxlinelength");
const QString MAXCPS_TAG = QStringLiteral("options.maxcps");
const QString MAXDURATION_TAG = QStringLiteral("options.maxduration");
const QString MINDURATION_TAG = QStringLiteral("options.minduration");
const QString SYSTEM_LASTFILE = QStringLiteral("system.lastfile");
const QString SCALING_TAG = QStringLiteral("ui.scaling.extra");
const QString OLD_SCALING_TAG = QStringLiteral("ui.scaling.factor");
const QString TIMESTAMP_TOOLTIPS_DISABLED = QStringLiteral("ui.tooltips.timestamp.disabled");
const QString USE_THEME_VARIATION = QStringLiteral("ui.theme.variation");
const QString LANGUAGE_TAG = QStringLiteral("ui.language");
const QString VIDEOPREVIEW_HARDWARE_TAG = QStringLiteral("videopreview.hardware");
const QString DEFAULT_ENCODING_8BIT_TAG = QStringLiteral("default.encoding.8bit");
const QString DEFAULT_ENCODING_CJK_TAG = QStringLiteral("default.encoding.cjk");

struct State {
    int errorColor = 1;
    bool spaceChars = false, newlineChars = false, compactSubs = true, otherChars = true;
    int maxLines = 2;
    float fillPercent = 50;
    int maxLineLength = 42;
    int maxCPS = 21;
    float maxDuration = 7, minDuration = 1;
    float scaling = 1.0f;
    bool timestampTooltipsDisabled = false;
    ThemeVariation themeVariation = ThemeVariation::AUTO;
    QString language = QStringLiteral("auto");
    bool videoPreviewHardware = false;
    QString defaultEncoding8bit = QStringLiteral("ISO-8859-1");
    QString defaultEncodingCjk;
    bool loaded = false;
};

State &st() {
    static State s;
    if (!s.loaded) {
        s.loaded = true;
        load();
    }
    return s;
}

// Up to 10.1 the preferences held "default.encoding1..3"; fold them into the
// 8-bit / CJK slots once.
void migrateDefaultEncoding() {
    if (!Prefs::getString(DEFAULT_ENCODING_8BIT_TAG, QString()).isNull())
        return;
    QString single, cjk;
    for (int i = 1; i <= 3; ++i) {
        const QString e = Prefs::getString(QStringLiteral("default.encoding%1").arg(i), QString());
        if (e.isNull() || isUnicodeCharset(e))
            continue;
        if (isSingleByteCharset(e)) {
            if (single.isNull()) single = e;
        } else if (cjk.isNull())
            cjk = e;
    }
    Prefs::set(DEFAULT_ENCODING_8BIT_TAG, single.isNull() ? QStringLiteral("ISO-8859-1") : single);
    if (!cjk.isNull())
        Prefs::set(DEFAULT_ENCODING_CJK_TAG, cjk);
}
}  // namespace

void load() {
    State &s = st();
    s.loaded = true;
    s.errorColor = Prefs::getInt(ERRORCOLOR_TAG, 1);
    s.spaceChars = Prefs::getBoolean(SPACECHARS_TAG, false);
    s.newlineChars = Prefs::getBoolean(NEWLINECHARS_TAG, false);
    s.compactSubs = Prefs::getBoolean(COMPACTSUBS_TAG, true);
    s.otherChars = Prefs::getBoolean(OTHERCHARS_TAG, true);
    s.maxLines = Prefs::getInt(MAXLINES_TAG, 2);
    s.fillPercent = Prefs::getFloat(FILLPERCENT_TAG, 50);
    s.maxLineLength = Prefs::getInt(MAXLINELENGTH_TAG, 42);
    s.maxCPS = Prefs::getInt(MAXCPS_TAG, 21);
    s.maxDuration = Prefs::getFloat(MAXDURATION_TAG, 7);
    s.minDuration = Prefs::getFloat(MINDURATION_TAG, 1);
    // The old total-scale key is abandoned (it double-scaled on systems that scale on their own).
    Prefs::remove(OLD_SCALING_TAG);
    s.scaling = Prefs::getFloat(SCALING_TAG, 1.0f);
    s.timestampTooltipsDisabled = Prefs::getBoolean(TIMESTAMP_TOOLTIPS_DISABLED, false);
    s.themeVariation = themeVariationFromName(Prefs::getString(USE_THEME_VARIATION, QStringLiteral("AUTO")));
    s.language = Prefs::getString(LANGUAGE_TAG, QStringLiteral("auto"));
    s.videoPreviewHardware = Prefs::getBoolean(VIDEOPREVIEW_HARDWARE_TAG, false);
    migrateDefaultEncoding();
    s.defaultEncoding8bit = Prefs::getString(DEFAULT_ENCODING_8BIT_TAG, QStringLiteral("ISO-8859-1"));
    if (!isSingleByteCharset(s.defaultEncoding8bit))  // the floor must always decode → single-byte
        s.defaultEncoding8bit = QStringLiteral("ISO-8859-1");
    s.defaultEncodingCjk = Prefs::getString(DEFAULT_ENCODING_CJK_TAG, QString());
}

bool isUnicodeCharset(const QString &name) {
    if (name.isNull()) return false;
    const QString u = name.toUpper();
    return u.startsWith(QLatin1String("UTF-")) || u.startsWith(QLatin1String("UTF8"))
        || u.startsWith(QLatin1String("UTF16")) || u.startsWith(QLatin1String("UTF32"))
        || u.startsWith(QLatin1String("X-UTF")) || u == QLatin1String("UTF");
}

bool isSingleByteCharset(const QString &name) {
    // Java: Charset.forName(name).newEncoder().maxBytesPerChar() == 1.
    return !isUnicodeCharset(name) && Charsets::isSingleByte(name);
}

#define OPT_SIMPLE(TYPE, GETTER, SETTER, FIELD, TAG) \
    TYPE GETTER() { return st().FIELD; }              \
    void SETTER(TYPE v) { Prefs::set(TAG, st().FIELD = v); }

OPT_SIMPLE(int, getErrorColor, setErrorColor, errorColor, ERRORCOLOR_TAG)
OPT_SIMPLE(bool, isSpaceChars, setSpaceChars, spaceChars, SPACECHARS_TAG)
OPT_SIMPLE(bool, isNewlineChars, setNewlineChars, newlineChars, NEWLINECHARS_TAG)
OPT_SIMPLE(bool, isCompactSubs, setCompactSubs, compactSubs, COMPACTSUBS_TAG)
OPT_SIMPLE(bool, isOtherChars, setOtherChars, otherChars, OTHERCHARS_TAG)
OPT_SIMPLE(int, getMaxLines, setMaxLines, maxLines, MAXLINES_TAG)
OPT_SIMPLE(int, getMaxLineLength, setMaxLineLength, maxLineLength, MAXLINELENGTH_TAG)
OPT_SIMPLE(int, getMaxCPS, setMaxCPS, maxCPS, MAXCPS_TAG)
OPT_SIMPLE(float, getMaxDuration, setMaxDuration, maxDuration, MAXDURATION_TAG)
OPT_SIMPLE(float, getMinDuration, setMinDuration, minDuration, MINDURATION_TAG)
OPT_SIMPLE(float, getScaling, setScaling, scaling, SCALING_TAG)
OPT_SIMPLE(bool, isTimestampTooltipsDisabled, setTimestampTooltipsDisabled, timestampTooltipsDisabled, TIMESTAMP_TOOLTIPS_DISABLED)
OPT_SIMPLE(bool, isVideoPreviewHardware, setVideoPreviewHardware, videoPreviewHardware, VIDEOPREVIEW_HARDWARE_TAG)

float getFillPercent() { return st().fillPercent; }
void setFillPercent(float value) {
    if (value < 0) value = 0;
    else if (value > 100) value = 100;
    Prefs::set(FILLPERCENT_TAG, st().fillPercent = value);
}

ThemeVariation getThemeVariation() { return st().themeVariation; }
void setThemeVariation(ThemeVariation v) {
    Prefs::set(USE_THEME_VARIATION, themeVariationName(v));
    st().themeVariation = v;
}

QString getLanguage() { return st().language; }
void setLanguage(const QString &lang) {
    Prefs::set(LANGUAGE_TAG, lang);
    st().language = lang;
}

QString getDefaultEncoding8bit() { return st().defaultEncoding8bit; }
QString getDefaultEncodingCjk() { return st().defaultEncodingCjk; }

void rememberEncoding(const QString &enc) {
    if (enc.isEmpty() || isUnicodeCharset(enc))
        return;
    if (isSingleByteCharset(enc)) {
        st().defaultEncoding8bit = enc;
        Prefs::set(DEFAULT_ENCODING_8BIT_TAG, enc);
    } else {
        st().defaultEncodingCjk = enc;
        Prefs::set(DEFAULT_ENCODING_CJK_TAG, enc);
    }
}

void saveFileList(const QList<SubFile> &recents) {
    int pos = recents.size();
    int counter = 0;
    while (pos > 0 && counter < MAX_RECENTS) {
        --pos;
        const SubFile &sfile = recents.at(pos);
        const QFileInfo f(sfile.getSaveFile());
        if (f.exists() && f.isFile()) {
            ++counter;
            Prefs::set(SYSTEM_LASTFILE + QString::number(counter), sfile.getPacked());
        }
    }
    while (counter < MAX_RECENTS)
        Prefs::remove(SYSTEM_LASTFILE + QString::number(++counter));
}

QList<SubFile> loadFileList() {
    QList<SubFile> files;
    for (int i = MAX_RECENTS; i > 0; --i) {
        SubFile sf;
        if (!SubFile::unpack(Prefs::getString(SYSTEM_LASTFILE + QString::number(i), QString()), sf))
            continue;
        const QFileInfo f(sf.getSaveFile());
        if (f.exists() && f.isReadable() && f.isFile())
            files.append(sf);
    }
    return files;
}

}  // namespace Options
